src/backend/distributed/commands/table.c

```diff
@@ -376,14 +379,31 @@ PreprocessAlterTableStmt(Node *node, const char *alterTableCommand)
 		leftRelationId = IndexGetRelation(leftRelationId, missingOk);
 	}
 
-	/*
-	 * Normally, we would do this check in ErrorIfUnsupportedForeignConstraintExists
-	 * in post process step. However, we skip doing error checks in post process if
-	 * this pre process returns NIL -and this method returns NIL if the left relation
-	 * is a postgres table. So, we need to error out for foreign keys from postgres
-	 * tables to citus local tables here.
-	 */
-	ErrorIfAlterTableDefinesFKeyFromPostgresToCitusLocalTable(alterTableStatement);
+	if (processUtilityContext != PROCESS_UTILITY_SUBCOMMAND &&
+		AlterTableDefinesFKeyBetweenPostgresAndNonDistTable(alterTableStatement))
+	{
```

Comment:
The following steps work on 9.5/master, but fail on this branch when the coordinator is not added to the metadata
```sql
create table ref (a int primary key);
select create_reference_table('ref');
create table other (x int primary key, y int);
alter table other add constraint fk foreign key (y) references ref (a) on delete cascade;
select create_reference_table('other');
```

```sql
create table ref (a int primary key);
select create_reference_table('ref');
create table other (x int primary key, y int);
alter table other add constraint fk foreign key (y) references ref (a) on delete cascade;
ERROR:  cannot create foreign key: could not find the coordinator node in metadata as it is not added as a worker
```

What happens if we skip this if when the coordinator is not in metadata?

Reply:
With the changes on #4489, the above seems to be covered. Still, the following is not:
```SQL
drop table if exists ref, ref2 cascade;


create table ref (a int primary key);
create table ref2 (x int);
alter table ref2 add constraint fk foreign key (x) references ref (a);
select create_reference_table('ref');
alter table ref2 add constraint fk2 foreign key (x) references ref (a);

ERROR:  cannot create foreign key: table "ref" is already distributed
Time: 2.276 ms
 ```
It looks like `create_reference_table` does not convert `ref2` to a citus local table, and then the second `ALTER TABLE` cascades from `ref2` to `ref` for creating citus local tables as where already exists a foreign key between them.

In an ideal world, we'd want `create_reference_table` to cascade referencing tables to local tables. However, given that we may skip cascade if the coordinator is not in the metadata, we could consider `CascadeOperationForConnectedRelations` to skip cascading into already `IsCitusTable` relations when the operation is `create_citus_local`

Reply:
I think we have two issues to fix here:

For the first `alter table ref2 add constraint` command, we should be converting postgres tables to citus local tables as we do in this pr.

For the second `alter table ref2 add constraint`, as @onderkalaci mentioned, we need to skip citus tables when calling `CreateCitusLocalTable`.

This is because if we only fix the second part, then even if the coordinator is added to metadata, then we won't be converting local tables to citus local tables for such `create_reference_table` calls, but seems that we should do that.

Reply:
I'm resolving this as  all the mentioned items are covered for both the coordinator added or not.